pkg/metadata/main.go

```diff
@@ -139,70 +139,80 @@ func main() {
 //        }
 // }
 // Will create foobar/foo with mode 0644 and content "hello"
-func processUserData(data []byte) error {
+func processUserData(basePath string, data []byte) error {
 	// Always write the raw data to a file
-	err := ioutil.WriteFile(path.Join(ConfigPath, "userdata"), data, 0644)
+	err := ioutil.WriteFile(path.Join(basePath, "userdata"), data, 0644)
 	if err != nil {
 		log.Printf("Could not write userdata: %s", err)
 		return err
 	}
 
-	var fd interface{}
-	if err := json.Unmarshal(data, &fd); err != nil {
+	var root map[string]entry
+	if err := json.Unmarshal(data, &root); err != nil {
 		// Userdata is no JSON, presumably...
 		log.Printf("Could not unmarshall userdata: %s", err)
 		// This is not an error
 		return nil
 	}
-	cm, ok := fd.(map[string]interface{})
-	if !ok {
-		log.Printf("Could convert JSON to desired format: %s", fd)
-		return nil
-	}
-	for d, val := range cm {
-		dir := path.Join(ConfigPath, d)
-		if err := os.MkdirAll(dir, 0755); err != nil {
-			log.Printf("Failed to create %s: %s", dir, err)
-			continue
-		}
-		files, ok := val.(map[string]interface{})
-		if !ok {
-			log.Printf("Could convert JSON for files: %s", val)
-			continue
+
+	var queue []entry
+	var current entry
+	queue = enqueue(queue, root, basePath)
+	for len(queue) != 0 {
+		current, queue = queue[0], queue[1:]
+		if current.Content != nil {
+			writeFile(current)
+		} else {
```

Comment:
I think we should check that exactly one of `Content` and `Entries` is non nil, i.e. that one but not both of them is set. That's reduce the possibility for confusion.

Reply:
Entries is, with this struct, never nil. I have to use `*map[string]entry` for that, which is a weird type, no ?

Reply:
I do the change for that also

Reply:
no, the empty map is `nil` but you have to initialize before writing. Go is weird. So you should be able to check if it is `nil` still...

Reply:
Sorry, I meant `null`/not present in the JSON sense, i.e. an empty map rather than a nil in the Go sense. (Is an uninitialised map == `nil` in Go? I'm not sure)

Reply:
(I hadn't seen Justin's comment yet when I wrote the above)

Reply:
yes, uninitialised map == `nil` thats what I was saying, you dont need a pointer

Reply:
we raced a second time ;-)